[masonry] Optimize items that span multiple tracks

stwrt opened the pull request:
#### f10662fe88d05adea16c6cdccd850905acd342f2
<pre>
[masonry] Optimize items that span multiple tracks
<a href="https://bugs.webkit.org/show_bug.cgi?id=276733">https://bugs.webkit.org/show_bug.cgi?id=276733</a>
<a href="https://rdar.apple.com/problem/132435056">rdar://problem/132435056</a>

Reviewed by Sammy Gill and Yusuke Suzuki.

The implementation is split up into several steps.

1. Gather indefinite items into groups, and calculate their min/max sizes. Definite items
should be documented for later use too.

2. Update intrinsic tracks with single span items.

3. Convert indefinite items to definite items.

3. Update instrinsic tracks with multi span items that do not cross a flex track.

4. Update flex tracks with items that are single or multi span that cross a flex track.

Note: This masonry implementation does not support Subgrid at this point.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
 (WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
 (WebCore::GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup):
 (WebCore::GridTrackSizingAlgorithm::itemSizeForTrackSizeComputationPhaseMasonry const):
 (WebCore::GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItemsMasonry):
 (WebCore::GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItemsMasonryWithFlex):
 (WebCore::GridTrackSizingAlgorithm::convertIndefiniteItemsToDefiniteMasonry):
 (WebCore::GridTrackSizingAlgorithm::distributeSpaceToTracks const):
 (WebCore::GridTrackSizingAlgorithmStrategy::minContentForGridItem const):
 (WebCore::GridTrackSizingAlgorithm::canParticipateInBaselineAlignment const):
 (WebCore::GridTrackSizingAlgorithm::computeIndefiniteItemsForMasonry):
 (WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry):
 (WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry): Deleted.
 (WebCore::GridTrackSizingAlgorithm::computeIndefiniteItemsForMasonry const): Deleted.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/282464@main">https://commits.webkit.org/282464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef2a81f2af268f59e9317c9394a1e7e7d959917a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42552 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15793 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67217 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13804 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65316 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14084 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9539 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66265 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54721 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31611 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36204 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12067 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12676 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68913 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7143 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12017 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58243 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7174 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54789 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14005 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5955 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38373 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39452 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39195 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WTF/wtf/StdMultiMap.h:
```diff
@@ -0,0 +1,38 @@
+namespace WTF {
+
+template<typename Key, typename Value, typename Compare = std::less<Key>, typename Allocator = FastAllocator<std::pair<const Key, Value>>>
+using StdMultiMap = std::multimap<Key, Value, Compare, Allocator>;
```
Based on discussion with Yusuke, make this `StdMultimap`.

stwrt commented on Source/ThirdParty/gtest/xcode/gtest.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 46;
```
Need to revert.

Constellation reviewed (approved):
r=me for WTF change.
But other part requires experts' review.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-002-expected.html:
```diff
@@ -283,20 +283,20 @@
   <item>3</item>
-  <item>4</item>
+  <item style="width:2ch">4</item>
   <item>5</item>
```
I think you need to leave the items alone here and just change the track sizing, right?

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-002-expected.html:
```diff
@@ -283,20 +283,20 @@
-  <item>4</item>
-  <item style="width:2ch">5</item>
+  <item style="width:2ch">4</item>
+  <item>5</item>
```
Same here.

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-002.html:
```diff
@@ -275,20 +275,20 @@
-  <item>4</item>
+  <item style="width:2ch;">4</item>
   <item>5</item>
 </grid>
```
Don't change the test here. Just the reference.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            track.setPlannedSize(trackSizeForTrackSizeComputationPhase(phase, track, TrackSizeRestriction::AllowInfinity));
+        }
+
+        Vector<WeakPtr<GridTrack>> growBeyondGrowthLimitsTracks;
+        Vector<WeakPtr<GridTrack>> filteredTracks;
```
Can probably use WeakListHashSet instead

stwrt replied:
This would require updating the grid implementation too, because the vectors are sent down to `distributeSpaceToTracks`.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
 }
 
+template <TrackSizeComputationVariant variant>
+void GridTrackSizingAlgorithm::masonryIncreaseSizesToAccommodateSpanningItems(const StdMap<unsigned, MasonryMinMaxTrackSize> gridTrackSpans, WTF::StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItems, GridLayoutState& gridLayoutState)
```
I can't really tell what the unsigned represents in this case. In track sizing this could probably refer to a track index, or span length so we should create an alias for it

sammygill replied:
`gridTrackSpans` doesn't seem right because a track doesn't have a span but an item does

stwrt replied:
This was totally redone with how the items are passed through the function.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+template <TrackSizeComputationVariant variant>
+void GridTrackSizingAlgorithm::masonryIncreaseSizesToAccommodateSpanningItems(const StdMap<unsigned, MasonryMinMaxTrackSize> gridTrackSpans, WTF::StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItems, GridLayoutState& gridLayoutState)
+{
+    auto increaseSizes = [&]<TrackSizeComputationPhase phase>(unsigned spanLength, WTF::StdMultimap<unsigned, MasonryMinMaxTrackSizeWithGridSpan> multiTrackItems)
```
Can we move this to a helper function defined outside? That would help improve the readability of the rest of this method

stwrt replied:
Discussed in person. I basically moved all the other code out of this function, so it is just the code to setup the call to the lambda.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            auto endLine = trackIndex + gridTrackSpan.first;
+            auto itemSpan = GridSpan::translatedDefiniteGridSpan(trackIndex, endLine);
+
+            if (itemSpan.integerSpan() == 1)
```
Since we handle this somewhere else we should have a comment that references that imo

stwrt replied:
Added a comment

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        }
+    }
+
+    for (auto& multiTrackItem : definiteMultiTrackItems) {
```
I don't think this should be `multiTrackItem` since we create a `multiTrackItems` vector above and this element is not from that vector

stwrt replied:
This was re-written.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+template <TrackSizeComputationVariant variant>
+void GridTrackSizingAlgorithm::masonryIncreaseSizesToAccommodateSpanningItemsWithFlex(const StdMap<unsigned, MasonryMinMaxTrackSize> gridTrackSpans, WTF::StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItems, GridLayoutState& gridLayoutState)
+{
+    auto increaseSizes = [&]<TrackSizeComputationPhase phase>(StdSet<unsigned> spans, WTF::StdMultimap<unsigned, MasonryMinMaxTrackSizeWithGridSpan> multiTrackItems)
```
Same thing here let's move it outside this method definition

stwrt replied:
Removed most of the other code out of this function. Please re-evaluate.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
-    Vector<GridItemWithSpan> itemsCrossingFlexibleTracks;
-    SingleThreadWeakHashSet<RenderBox> itemsSet;
-    MasonryIndefiniteItems masonryIndefiniteItems;
+    StdMap<unsigned, MasonryMinMaxTrackSize> indefiniteTrackSizes;
```
The value here seems to come from the intrinsic sizes of grid items for a particular span length and I don't think `indefiniteTrackSizes` is that

stwrt replied:
Updated the name

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
-    if (m_grid.hasGridItems()) {
-        computeIndefiniteItemsForMasonry(masonryIndefiniteItems, gridLayoutState);
+    // Compute Span Sizes
+    computeIndefiniteItemsForMasonry(indefiniteTrackSizes, definiteMultiTrackItems, definiteMultiTrackItemsWithFlex, gridLayoutState);
```
We should consider if we can try to retrun something here by value instead I think that would make it a bit cleaner

stwrt replied:
This would need to return 3 different items 🤔

stwrt replied:
We can create a struct or another form to hold the return, but the current approach feels simpler to me.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
+    auto& allTracks = tracks(m_direction);
 
+    if (indefiniteTrackSizes.contains(1)) {
+        auto& singleTrackSpanSize = indefiniteTrackSizes.find(1)->second;
```
You can probably rewrite this to avoid the redudant lookups by moving the find into the conditional

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
-    }
-    GridItemsSpanGroupRange tracksGroupRange = { itemsCrossingFlexibleTracks.begin(), itemsCrossingFlexibleTracks.end() };
-    increaseSizesToAccommodateSpanningItems<TrackSizeComputationVariant::CrossingFlexibleTracks>(tracksGroupRange, gridLayoutState);
+    // Multi Tracks (w/o Flex) Spans
```
```suggestion
    // Multi Tracks (w/o Flex) Spans
```
// Multi Tracks without Flex Spans.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
+    // Multi Tracks (w/o Flex) Spans
+    masonryIncreaseSizesToAccommodateSpanningItems<TrackSizeComputationVariant::NotCrossingFlexibleTracks>(indefiniteTrackSizes, definiteMultiTrackItems, gridLayoutState);
+
+    // Multi Tracks (w/ Flex) Spans
```
```suggestion
    // Multi Tracks (w/ Flex) Spans
```// Multi Tracks with Flex Spans.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1815,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
 
-            accumulateIntrinsicSizesForTrackMasonry(track, trackIndex, iterator, itemsSortedByIncreasingSpan, itemsCrossingFlexibleTracks, itemsSet, masonryIndefiniteItems, 0_lu, gridLayoutState);
+            auto itemSpan = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + 1);
+            if (spanningItemCrossesFlexibleSizedTracks(itemSpan))
+                continue;
```
Is it possible for this to actually happen if we are iterating over `m_contentSizedTracksIndex` with the items we are considering? AFAICT this only gets populated if the corresponding track size is content sized.

stwrt replied:
Just tested this out. We do get flex tracks in here.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,22 +273,22 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
 }
 
-void GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup(const GridSpan& span, MasonryIndefiniteItems& masonryIndefiniteItems, GridTrack& track)
+void GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup(const GridSpan& span, MasonryMinMaxTrackSize& spanSize, GridTrack& track)
```
A span size, which is a numeric value, does not have min/max content sizes which are associated with grid items

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -327,6 +330,18 @@ bool GridTrackSizingAlgorithm::spanningItemCrossesFlexibleSizedTracks(const Grid
     return false;
 }
 
+bool GridTrackSizingAlgorithm::spanningItemCrossesIntrinsicSizedTracks(const GridSpan& itemSpan) const
```
Did you end up using this method? I did a quick search and didn't see it. If you did end up using it can we reuse `GridTrackSizingAlgorithm::isIntrinsicSizedGridArea` for your use cases?

stwrt replied:
When I did a rewrite I removed all uses of it; removing now.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+    };
+
+    StdSet<unsigned> spanLengths;
+    WTF::StdMultimap<unsigned, MasonryMinMaxTrackSizeWithGridSpan> multiTrackItems;
```
If we wanted to follow naming closer to the spec we could rename this to something like `itemsSpanningMultipleTracks`

stwrt replied:
This might be a better name. Will continue to experiment.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            if (spanningItemCrossesFlexibleSizedTracks(itemSpan))
+                continue;
+
+            multiTrackItems.insert({ itemSpan.integerSpan(), MasonryMinMaxTrackSizeWithGridSpan { .trackSize = gridTrackSpan.second, .gridSpan = itemSpan } });
```
```suggestion
            multiTrackItems.insert({ itemSpan.integerSpan(), MasonryMinMaxTrackSizeWithGridSpan { gridTrackSpan.second, itemSpan } });
```

stwrt replied:
It seems a bit confusing not knowing the type you are initiating, but I also need to verify the WebKit style guide here.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+
+            multiTrackItems.insert({ itemSpan.integerSpan(), MasonryMinMaxTrackSizeWithGridSpan { .trackSize = gridTrackSpan.second, .gridSpan = itemSpan } });
+
+            if (!spanLengths.contains(itemSpan.integerSpan()))
```
Since `spanLengths` is a set can we just unconditionally `insert` into it rather than checking if it does not contain the unsigned?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        multiTrackItems.insert({ multiTrackItem.first, MasonryMinMaxTrackSizeWithGridSpan { .trackSize = MasonryMinMaxTrackSize { .minContentSize = minContentForGridItem, .maxContentSize = maxContentForGridItem, .minSize = minSizeForGridItem }, .gridSpan = multiTrackItem.second.span() } });
+
+        if (!spanLengths.contains(multiTrackItem.first))
+            spanLengths.insert(multiTrackItem.first);
```
Same comment about inserting into a set here

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        auto maxContentForGridItem = m_strategy->maxContentForGridItem(gridItem, gridLayoutState);
+        auto minSizeForGridItem = m_strategy->minSizeForGridItem(gridItem, gridLayoutState);
+
+        multiTrackItems.insert({ multiTrackItem.first, MasonryMinMaxTrackSizeWithGridSpan { .trackSize = MasonryMinMaxTrackSize { .minContentSize = minContentForGridItem, .maxContentSize = maxContentForGridItem, .minSize = minSizeForGridItem }, .gridSpan = multiTrackItem.second.span() } });
```
This is fairly hard to parse by just glancing over the code since we're creating a couple of different things. I think we should consider refactoring this and put each of the fields of the`MasonryMinMaxTrackSizeWithGridSpan` into their own locals and use those when inserting into the track.

stwrt replied:
This piece of code is very terse; I can look into refactoring it.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            track.setPlannedSize(trackSizeForTrackSizeComputationPhase(phase, track, TrackSizeRestriction::AllowInfinity));
+        }
+
+        Vector<WeakPtr<GridTrack>> growBeyondGrowthLimitsTracks;
+        Vector<WeakPtr<GridTrack>> filteredTracks;
```
We should probably use `WeakListHashSet`

stwrt replied:
As mentioned in another comment this will require updating the grid implementation too.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        auto maxContentForGridItem = m_strategy->maxContentForGridItem(gridItem, gridLayoutState);
+        auto minSizeForGridItem = m_strategy->minSizeForGridItem(gridItem, gridLayoutState);
+
+        multiTrackItems.insert({ multiTrackItem.first, MasonryMinMaxTrackSizeWithGridSpan { .trackSize = MasonryMinMaxTrackSize { .minContentSize = minContentForGridItem, .maxContentSize = maxContentForGridItem, .minSize = minSizeForGridItem }, .gridSpan = multiTrackItem.second.span() } });
```
Same thing here about renaming these fields for clarity

stwrt replied:
Same as comment above.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        multiTrackItems.insert({ multiTrackItem.first, MasonryMinMaxTrackSizeWithGridSpan { .trackSize = MasonryMinMaxTrackSize { .minContentSize = minContentForGridItem, .maxContentSize = maxContentForGridItem, .minSize = minSizeForGridItem }, .gridSpan = multiTrackItem.second.span() } });
+
+        if (!spanLengths.contains(multiTrackItem.first))
+            spanLengths.insert(multiTrackItem.first);
```
Probably just `insert`?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +574,204 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+    auto& allTracks = tracks(m_direction);
+
+    for (auto& gridTrackSpan : gridTrackSpans) {
+        for (auto i = 0u; i < allTracks.size(); i++) {
```
I think we can come up with a better name than just `i` :)

stwrt replied:
Fixed

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
```diff
@@ -163,28 +166,17 @@ class GridTrackSizingAlgorithm final {
+    struct MasonryMinMaxTrackSizeWithGridSpan {
+        MasonryMinMaxTrackSize trackSize;
+        GridSpan gridSpan;
     };
```
Spacing between these definitions would be nice imo

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1758,43 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
-    SingleThreadWeakHashSet<RenderBox> itemsSet;
-    MasonryIndefiniteItems masonryIndefiniteItems;
+    // Compute Span Sizes.
+    computeIndefiniteItemsForMasonry(indefiniteItems, definiteItems, definiteItemsThatSpanFlexTrack, gridLayoutState);
```
Do we want to consider having `computeIndefiniteItemsForMasonry` return a struct that contains your three data structures instead?

sammygill replied:
Also, it's a little strange now that this method's name is `computeIndefiniteItemsForMasonry` but you pass in data structures that seem to hold definite items also to populate. I think either this method's name needs to change or we need to refactor the logic

stwrt replied:
Good point. I need to update the name for this; it held true at one point, but is no longer a good name.

stwrt replied:
Updated to `computeDefiniteAndIndefiniteItemsForMasonry`. This better identifies that we are populating the both indefinite and definite items for later use.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1758,43 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
+    auto& allTracks = tracks(m_direction);
 
+    auto item = indefiniteItems.find(1);
+    if (item != indefiniteItems.end()) {
```
Since you don't use `item` outside of the if

```suggestion    
    if (auto item = indefiniteItems.find(1); item != indefiniteItems.end()) {
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1653,63 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-        if (renderGrid()->isMasonry() && GridPositionsResolver::resolveGridPositionsFromStyle(*m_renderGrid, *gridItem.get(), m_direction).isIndefinite())
-            continue;
+    auto& allTracks = tracks(m_direction);
+    auto trackLength = allTracks.size();
```
```suggestion
    auto tracksCount = allTracks.size();
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1653,63 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-            accumulateIntrinsicSizes(&gridItem);
-    }
-}
+            // Definite items need to be handled individually, and will not be included in the global track sizing.
```
It's hard to understand what this comment is saying. It implies there is a "global," and "non global," track sizing but it's not clear what those mean

stwrt replied:
Yeah, this comment isn't super helpful anymore. Deleting it.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1653,63 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        else if (spanLength == 1)
+            sizeTrackToFitNonSpanningItem(gridSpan, *gridItem, allTracks[trackIndex], gridLayoutState);
+        else
+            definiteItems.insert({ spanLength, MasonryMinMaxTrackSizeWithGridSpan { .trackSize = MasonryMinMaxTrackSize { .minContentSize = minContentForGridItem, .maxContentSize = maxContentForGridItem, .minSize = minSizeForGridItem }, .gridSpan = gridSpan } });
```
I really think we should refactor this code to improve the readability of it. Maybe we can rework this logic so that a conditional for single track items not inside a flexible track should be the first case. Aftewards you can create your `MasonryMinMaxTrackSizeWithGridSpan` and `MasonryMinMaxTrackSize` objects in named variables that can by used instead when you add to your structures.

stwrt replied:
I am think this works pretty well.

```
bool spansFlexTracks = spanningItemCrossesFlexibleSizedTracks(gridSpan);

if (spanLength == 1 && !spansFlexTracks)
    sizeTrackToFitNonSpanningItem(gridSpan, *gridItem, allTracks[trackIndex], gridLayoutState);
else {
    auto minMaxTrackSizeWithGridSpan = MasonryMinMaxTrackSizeWithGridSpan { MasonryMinMaxTrackSize { minContentForGridItem, maxContentForGridItem, minSizeForGridItem }, gridSpan };

    if (spansFlexTracks)
        definiteItemsThatSpanFlexTrack.append(minMaxTrackSizeWithGridSpan);
    else
        definiteItems.insert({ spanLength, minMaxTrackSizeWithGridSpan });
}
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1758,43 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
+    // Update intrinsic tracks with mutli span items that do not cross flex tracks.
+    increaseSizesToAccommodateSpanningItemsMasonry<TrackSizeComputationVariant::NotCrossingFlexibleTracks>(definiteItems);
+
+    // Update flex tracks with single and mutli span items that cross flex tracks.
```
Not really sure these comments are adding anything because they are basically just repeating what the code is doing. For the first comment it would be better to explain *why* we need to convert these items if that isn't clear.

stwrt replied:
Removed the comments.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,22 +273,22 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
-        auto growthLimit = masonryIndefiniteItems.largestMaxContentSizeForSingleTrackItems;
+        auto growthLimit = itemSize.maxContentSize;
         if (trackSize.isFitContent())
             growthLimit = std::min(growthLimit, valueForLength(trackSize.fitContentTrackBreadth().length(), availableSpace().value_or(0)));
```
Are the previous names that were used here no longer accurate? They seemed to be fairly descriptive of what they wer eat the time and it doesn't seem like this is the size for an item?

stwrt replied:
Updated with a revert back to the original name.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        increaseSizes.template operator()<TrackSizeComputationPhase::ResolveContentBasedMinimums>(span, definiteItems);
+        increaseSizes.template operator()<TrackSizeComputationPhase::ResolveMaxContentMinimums>(span, definiteItems);
+        increaseSizes.template operator()<TrackSizeComputationPhase::ResolveIntrinsicMaximums>(span, definiteItems);
+        increaseSizes.template operator()<TrackSizeComputationPhase::ResolveMaxContentMaximums>(span, definiteItems);
```
Can we use a structured binding instead of creating this temporary `StdSet` just to store the keys? Can we do something like `for (auto[spanSize, largestIntrinsicSizes] : definiteItems)` work?

stwrt replied:
I can refactor this to how Grid is doing it.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+{
+    auto increaseSizes = [&]<TrackSizeComputationPhase phase>(Vector<MasonryMinMaxTrackSizeWithGridSpan>& definiteItems)
+    {
+        Vector<GridTrack>& allTracks = tracks(m_direction);
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        Vector<WeakPtr<GridTrack>> filteredTracks;
+
+        for (auto& item : definiteItems) {
+            const GridSpan& itemSpan = item.gridSpan;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            growBeyondGrowthLimitsTracks.shrink(0);
+            LayoutUnit spanningTracksSize;
+            for (auto trackPosition : itemSpan) {
+                GridTrack& track = allTracks[trackPosition];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+
+            spanningTracksSize += m_renderGrid->guttersSize(m_direction, itemSpan.startLine(), itemSpan.integerSpan(), availableSpace());
+
+            LayoutUnit extraSpace = itemSizeForTrackSizeComputationPhaseMasonry(phase, item.trackSize) - spanningTracksSize;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        }
+
+        for (const auto& trackIndex : m_contentSizedTracksIndex) {
+            GridTrack& track = allTracks[trackIndex];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+    {
+        Vector<GridTrack>& allTracks = tracks(m_direction);
+        for (const auto& trackIndex : m_contentSizedTracksIndex) {
+            GridTrack& track = allTracks[trackIndex];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        }
+
+        for (const auto& trackIndex : m_contentSizedTracksIndex) {
+            GridTrack& track = allTracks[trackIndex];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+
+            spanningTracksSize += m_renderGrid->guttersSize(m_direction, itemSpan.startLine(), itemSpan.integerSpan(), availableSpace());
+
+            LayoutUnit extraSpace = itemSizeForTrackSizeComputationPhaseMasonry(phase, it->second.trackSize) - spanningTracksSize;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+            growBeyondGrowthLimitsTracks.shrink(0);
+            LayoutUnit spanningTracksSize;
+            for (auto trackPosition : itemSpan) {
+                GridTrack& track = allTracks[trackPosition];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+        Vector<WeakPtr<GridTrack>> filteredTracks;
+
+        for (auto [it, rangeEnd] = definiteItems.equal_range(spanLength); it != rangeEnd; it++) {
+            const GridSpan& itemSpan = it->second.gridSpan;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -539,6 +562,155 @@ void GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItems(const Gri
+    {
+        Vector<GridTrack>& allTracks = tracks(m_direction);
+        for (const auto& trackIndex : m_contentSizedTracksIndex) {
+            GridTrack& track = allTracks[trackIndex];
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1653,63 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
 
-        if (!isNewEntry)
-            return;
+        MasonryMinMaxTrackSize& trackSize = indefiniteItems.find(spanLength)->second;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1649,68 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-    auto accumulateIntrinsicSizes = [&](RenderBox* gridItem) {
-        bool isNewEntry = itemsSet.add(*gridItem).isNewEntry;
-        GridSpan span = m_renderGrid->gridSpanForGridItem(*gridItem, m_direction);
+    auto populateDefiniteItems = [&](unsigned trackIndex, GridSpan& gridSpan, unsigned spanLength, CheckedPtr<RenderBox> gridItem, Vector<GridTrack> & allTracks) {
```
You already protect `gridItem` on the stack so this should not be a `CheckedPtr` and should just be a raw pointer.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1481,116 +1649,68 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-        if (CheckedPtr inner = dynamicDowncast<RenderGrid>(gridItem); inner && inner->isSubgridInParentDirection(iterator.direction())) {
-            // Contribute the mbp of wrapper to the first and last tracks that we span.
-            GridSpan subgridSpan = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + span.integerSpan());
+    auto populateIndefiniteItems = [&](CheckedPtr<RenderBox> gridItem, unsigned spanLength) {
```
You already protect `gridItem` on the stack so this should not be a `CheckedPtr` and should just be a raw pointer.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1759,39 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
         return;
     }
+    StdMap<SpanLength, MasonryMinMaxTrackSize> indefiniteItems;
+    StdMap<SpanLength, Vector<MasonryMinMaxTrackSizeWithGridSpan>> definiteItems;
```
We should probably indicate that these are indefinite items spanning flex tracks, as opposed to its counterpart, so maybe something like `indefiniteItemsSpanningFlexTracks`?

stwrt replied:
That is not quite correct. `indefiniteItems` does not care whether we span flex tracks or not. When we "create fake grid items" in `convertIndefiniteItemsToDefiniteMasonry` we start to care about whether it spans or flex tracks or not, but the results of that action are stored in the definite items.

sammygill replied:
Oops I meant the `definiteItems` structure. Does that still hold up?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1639,38 +1759,39 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
     }
+    StdMap<SpanLength, MasonryMinMaxTrackSize> indefiniteItems;
+    StdMap<SpanLength, Vector<MasonryMinMaxTrackSizeWithGridSpan>> definiteItems;
+    Vector<MasonryMinMaxTrackSizeWithGridSpan> definiteItemsThatSpanFlexTrack;
```
Nit: I feel like "That," in this name doesn't really add anything so this could probably be called `definiteItemsSpanningFlexTracks`

stwrt replied:
That sounds a bit better; will update.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1496,116 +1664,68 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
 
-        accumulateIntrinsicSizes(gridItem.get());
-    }
+        while (CheckedPtr<RenderBox> gridItem = iterator.nextGridItem()) {
```
I think this should just be `CheckedPtr`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1496,116 +1664,68 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-            GridIterator subgridIterator = GridIterator::createForSubgrid(*inner, iterator, subgridSpan);
-            MasonryIndefiniteItems subgridMasonryIndefiniteItems;
+        if (!indefiniteItems.contains(spanLength))
+            indefiniteItems.insert({ spanLength, MasonryMinMaxTrackSize { 0, 0, 0 } });
```
Optionally can write this as `indefiniteItems.insert({ spanLength, { } });`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1654,38 +1774,38 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
     }
+    StdMap<SpanLength, MasonryMinMaxTrackSize> indefiniteItems;
+    StdMap<SpanLength, Vector<MasonryMinMaxTrackSizeWithGridSpan>> definiteItems;
+    Vector<MasonryMinMaxTrackSizeWithGridSpan> definiteItemsSpanFlexTrack;
```
Personally not a fan of how these structures are named as holding "items," but they don't actually hold items but instead sizes computed in a certain context

stwrt committed:
[masonry] Optimize items that span multiple tracks
https://bugs.webkit.org/show_bug.cgi?id=276733
rdar://problem/132435056

Reviewed by Sammy Gill and Yusuke Suzuki.

The implementation is split up into several steps.

1. Gather indefinite items into groups, and calculate their min/max sizes. Definite items
should be documented for later use too.

2. Update intrinsic tracks with single span items.

3. Convert indefinite items to definite items.

3. Update instrinsic tracks with multi span items that do not cross a flex track.

4. Update flex tracks with items that are single or multi span that cross a flex track.

Note: This masonry implementation does not support Subgrid at this point.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
 (WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
 (WebCore::GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup):
 (WebCore::GridTrackSizingAlgorithm::itemSizeForTrackSizeComputationPhaseMasonry const):
 (WebCore::GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItemsMasonry):
 (WebCore::GridTrackSizingAlgorithm::increaseSizesToAccommodateSpanningItemsMasonryWithFlex):
 (WebCore::GridTrackSizingAlgorithm::convertIndefiniteItemsToDefiniteMasonry):
 (WebCore::GridTrackSizingAlgorithm::distributeSpaceToTracks const):
 (WebCore::GridTrackSizingAlgorithmStrategy::minContentForGridItem const):
 (WebCore::GridTrackSizingAlgorithm::canParticipateInBaselineAlignment const):
 (WebCore::GridTrackSizingAlgorithm::computeIndefiniteItemsForMasonry):
 (WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry):
 (WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrackMasonry): Deleted.
 (WebCore::GridTrackSizingAlgorithm::computeIndefiniteItemsForMasonry const): Deleted.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/282464@main
